File: src/main/Java/boj_step20/boj20920.java
import java.io.*;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;
import java.util.stream.Stream;

public class boj20920 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        int n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());
        HashMap<String, Integer> notes = new HashMap<>();
        while (n-- > 0) {
            String word = reader.readLine();
            if (word.length() >= m) {
                if (notes.containsKey(word)) notes.put(word, notes.get(word) + 1);
                else notes.put(word, 1);
            }
        }
        Stream<Map.Entry<String, Integer>> result = notes.entrySet().stream().sorted((x, y) -> {
            if (x.getValue() != y.getValue()) return y.getValue().compareTo(x.getValue());
            else if (y.getKey().length() != x.getKey().length()) return y.getKey().length() - x.getKey().length();
            else return x.getKey().compareTo(y.getKey());
        });
        StringBuilder tmp = new StringBuilder();
        result.forEach(x -> tmp.append(x.getKey() + "\n"));
        writer.append(tmp);
        writer.close();
    }
}

File: src/main/Kotlin/boj_step20/boj20920.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val (n, m) = r.readLine().split(" ").map { it.toInt() }
    val notes = HashMap<String, Int>()
    repeat(n) {
        val word = r.readLine()
        if (word.length >= m) {
            if (word !in notes) notes.put(word, 1)
            else notes.set(word, notes.getValue(word) + 1)
        }
    }
    val result = notes.entries.sortedWith({ x, y ->
        when {
            x.value != y.value -> {
                y.value.compareTo(x.value)
            }
            x.key.length != y.key.length -> {
                y.key.length.compareTo(x.key.length)
            }
            else -> {
                x.key.compareTo(y.key)
            }
        }
    })
    result.forEach { w.write("${it.key}\n") }
    w.close()
}